### PR TITLE
Make "detected_host_os()" available on the SLAVE side of the split keyboard

### PR DESCRIPTION
**Pull request by ngs-ch**
## Description

Last year, "detected_host_os()" was merged into develop branche. However, it was only available on the master side for split keyboards.
So, I have changed it to synchronize with or without #define as well as the existing "Data Sync Options".

### Hwo to use
In `config.h` file.
`#define SPLIT_DETECTED_OS_ENABLE`

Note: We have confirmed that an error occurs if firmware is not written to both the left and right side when activating for the first time.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by daskygit** [comment]
I don't think this is the correct way to go about implementing this. There's already a function for retrieving the detected os that could be used, adding a further function that allows you to manually override/set the detected os may be better.

My other concern is does this need implementing at all in core?

This can be synced using the custom data sync option and with the flaky nature of os detection it's a feature that may end up not being used or removed.

**Comment by ngs-ch** [comment]
Thank you for your comment.

> There's already a function for retrieving the detected os that could be used

It seems to be parsing USB connection information and that function is not working on the slave side.
Could it be that I am using it wrong?

> My other concern is does this need implementing at all in core?
This can be synced using the custom data sync option

This is because the data is generated by core and is not touched by the user.
we can certainly do this with the custom data sync option. But, having said that, wouldn't the existing options do that as well? (e.g., SPLIT_LAYER_STATE_ENABLE)

>  os detection it's a feature that may end up not being used or removed.

I think it should be removed with them at that time.


I use a translation tool, so I apologize if there are any strange expressions.

File: quantum/os_detection.h [hunk 1]
```diff
@@ -27,6 +27,11 @@ typedef enum {
     OS_IOS,
 } os_variant_t;
 
+/* splits where the slave side needs to process for rgb/oled functionality
+ * define "SPLIT_DETECTED_OS_ENABLE" to enable  
+ */
+extern os_variant_t detected_os;
```
Comment (drashna) [inline quantum/os_detection.h:33]: This really shouldn't be exposed, TBH. 

Ideally should have a function to update it on the slave side. Eg: 
```c
#    if defined(SPLIT_KEYBOARD) && defined(SPLIT_DETECTED_OS_ENABLE)
void slave_update_detected_host_os(os_variant_t state) {
    detected_os = state;
}
#    endif
```

File: quantum/split_common/transaction_id_define.h [hunk 2]
```diff
@@ -109,6 +109,10 @@ enum serial_transaction_id {
     SPLIT_TRANSACTION_IDS_USER,
 #endif // SPLIT_TRANSACTION_IDS_USER
 
+#if defined(SPLIT_DETECTED_OS_ENABLE)
```
Comment (drashna) [inline quantum/split_common/transaction_id_define.h:112]: Also, this should go before the kb+users transactions, I think. 
```suggestion
#if defined(OS_DETECTION_ENABLE) && defined(SPLIT_DETECTED_OS_ENABLE)
```

File: quantum/split_common/transactions.c [hunk 3]
```diff
@@ -790,6 +790,33 @@ static void haptic_handlers_slave(matrix_row_t master_matrix[], matrix_row_t sla
 
 #endif // defined(HAPTIC_ENABLE) && defined(SPLIT_HAPTIC_ENABLE)
 
+////////////////////////////////////////////////////
+// Detected OS
+
+#if defined(SPLIT_DETECTED_OS_ENABLE)
```
Comment (drashna) [inline quantum/split_common/transactions.c:796]:
```suggestion
#if defined (OS_DETECTION_ENABLE) && defined(SPLIT_DETECTED_OS_ENABLE)
```

File: quantum/split_common/transport.h [hunk 4]
```diff
@@ -133,6 +133,10 @@ typedef struct _rpc_sync_info_t {
 } rpc_sync_info_t;
 #endif // defined(SPLIT_TRANSACTION_IDS_KB) || defined(SPLIT_TRANSACTION_IDS_USER)
 
+#if defined(SPLIT_DETECTED_OS_ENABLE)
```
Comment (drashna) [inline quantum/split_common/transport.h:136]:
```suggestion
#if defined(OS_DETECTION_ENABLE) && defined(SPLIT_DETECTED_OS_ENABLE)
```

File: quantum/split_common/transport.h [hunk 5]
```diff
@@ -209,6 +213,10 @@ typedef struct _split_shared_memory_t {
     uint8_t         rpc_m2s_buffer[RPC_M2S_BUFFER_SIZE];
     uint8_t         rpc_s2m_buffer[RPC_S2M_BUFFER_SIZE];
 #endif // defined(SPLIT_TRANSACTION_IDS_KB) || defined(SPLIT_TRANSACTION_IDS_USER)
+
+#if defined(SPLIT_DETECTED_OS_ENABLE)
```
Comment (drashna) [inline quantum/split_common/transport.h:217]:
```suggestion
#if defined(OS_DETECTION_ENABLE) && defined(SPLIT_DETECTED_OS_ENABLE)
```

**Comment by ngs-ch** [comment]
Thank you for pointing this out.
I have fixed it.

**Review by drashna — COMMENTED** [review]
Could you run a format pass on these files?

**Review by drashna — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**